pageserver/src/tenant/timeline/layer_manager.rs

```diff
@@ -390,42 +435,53 @@ impl LayerManagerWriteGuard {
     }
 
     /// Called when compaction is completed.
-    pub(crate) async fn finish_compact_l0(
-        &mut self,
-        layer_removal_cs: Arc<tokio::sync::OwnedMutexGuard<()>>,
+    pub(crate) async fn finish_compact_l0_consume_guard(
+        self,
+        _layer_removal_cs: Arc<tokio::sync::OwnedMutexGuard<()>>,
         compact_from: Vec<Arc<dyn PersistentLayer>>,
         compact_to: Vec<Arc<dyn PersistentLayer>>,
         metrics: &TimelineMetrics,
     ) -> Result<()> {
-        self.layer_manager
+        assert!(
+            self.pseudo_lock.is_right(),
+            "should use `modify` guard for this function."
+        );
+        let compact_from = self
+            .layer_manager
             .update(|mut layer_map| {
                 let mut updates = layer_map.batch_update();
                 for l in compact_to {
                     Self::insert_historic_layer(l, &mut updates, &self.snapshot.layer_fmgr);
                 }
-                for l in compact_from {
-                    // NB: the layer file identified by descriptor `l` is guaranteed to be present
-                    // in the LayerFileManager because compaction kept holding `layer_removal_cs` the entire
-                    // time, even though we dropped `Timeline::layers` inbetween.
-                    if let Err(e) = Self::delete_historic_layer(
-                        layer_removal_cs.clone(),
-                        l,
-                        &mut updates,
-                        metrics,
-                        &self.snapshot.layer_fmgr,
-                    ) {
-                        // If this fails, we will need to return the "partially" modified layer map
-                        // now. Eventually, we should decouple file deletion and layer map updates, so
-                        // that this part can be moved out of the `update` section.
-                        updates.flush();
-                        return Ok((layer_map, Err(e)));
-                    }
+                for l in &compact_from {
+                    // only remove from the layer map, not from file manager
+                    updates.remove_historic(l.layer_desc().clone());
                 }
                 updates.flush();
-                Ok((layer_map, Ok(())))
+                Ok((layer_map, Ok::<_, anyhow::Error>(compact_from)))
             })
-            .await
-            .unwrap() // unwrap the first level error, which is always Ok.
+            .await??;
+        drop(self.pseudo_lock);
+        // acquire the write lock so that all read threads are blocked, and once this lock is acquired,
```

Comment:
The concept looks quite similar to rcu family primitives. Ref https://github.com/neondatabase/neon/blob/main/libs/utils/src/simple_rcu.rs

Not asking to use it, just for reference